katcp/ioloop_manager.py

```diff
@@ -158,3 +165,84 @@ def join(self, timeout=None):
             self._ioloop_thread.join(timeout)
         except AttributeError:
             raise RuntimeError('Cannot join if not started')
+
+class IOLoopThreadWrapper(object):
+    default_timeout = None
+
+    def __init__(self, ioloop=None):
+        self.ioloop = ioloop = ioloop or tornado.ioloop.IOLoop.current()
+        self._thread_id = None
+        ioloop.add_callback(self._install)
+
+    def call_in_ioloop(self, fn, args, kwargs, timeout=None):
+        timeout = timeout or self.default_timeout
+        if get_thread_ident() == self._thread_id:
+            raise RuntimeError("Cannot call a thread-wrapped object from the ioloop")
+        future, tornado_future = Future(), tornado_Future()
+        self.ioloop.add_callback(
+            self._ioloop_call, future, tornado_future, fn, args, kwargs)
+        try:
+            # Use the threadsafe future to block
+            return future.result(timeout)
+        except TimeoutError:
+            raise
+        except Exception:
+            # If we have an exception use the tornado future instead since it
+            # will print a nicer traceback.
+            tornado_future.result()
+            # Should never get here since the tornado future should raise
+            assert False, 'Tornado Future should have raised'
+
+    def decorate_callable(self, callable_):
+        """Decorate a callable to use call_in_ioloop"""
+        @wraps(callable_)
+        def decorated(*args, **kwargs):
+            # Extract timeout from request itself or use default for ioloop wrapper
+            timeout = kwargs.get('timeout')
```

Comment:
I just moved the class, I wasn't planning on opening a can of worms ;)

I think the rationale is that the most likely reason for adding a timeout argument would be to support the ioloop-bouncy-timeout. If you then leave in the kwarg for an original callable that does not take a `timeout` kwarg you would get an exception. Perhaps it would make sense to rename the kwarg to something less likely to conflict with the original?